script/v2/manual-runs/scWETHv2Rebalance.s.sol

```diff
@@ -0,0 +1,392 @@
+    //         allocationPercents.push(COMPOUNDV3_ALLOCATION_PERCENT);
+    //         targetLtv[compoundV3Adapter] = COMPOUNDV3_TARGET_LTV;
+
+    //         totalAllocationPercent += COMPOUNDV3_ALLOCATION_PERCENT;
+    //     }
+
+    //     require(totalAllocationPercent == 1e18, "totalAllocationPercent != 100%");
+    // }
+
+    struct RebalanceDataParams {
+        IAdapter adapter;
+        uint256 amount;
+        uint256 flashLoanAmount;
+        bool isSupplyBorrow;
+    }
+
+    RebalanceDataParams[] rebalanceDataParams;
+
+    function _investAmount() internal view returns (uint256 investAmount) {
+        uint256 float = weth.balanceOf(address(vault));
+        uint256 minimumFloat = vault.minimumFloatAmount();
+        // investAmount == 0 just reinvests profits
+        investAmount = float > minimumFloat ? float - minimumFloat : 0;
+    }
+
+    function _updateRebalanceParams() internal {
+        uint256 investAmount = _investAmount();
+
+        IAdapter[3] memory allAdapters = [morphoAdapter, compoundV3Adapter, aaveV3Adapter];
+
+        IAdapter adapter;
+        uint256 allocationPercent;
+        uint256 amount;
+        uint256 flashLoanAmount;
+        bool isSupplyBorrow;
+        for (uint256 i; i < allAdapters.length; i++) {
+            adapter = allAdapters[i];
+            allocationPercent = adapterAllocationPercent[adapter];
+
+            if (allocationPercent != 0) {
+                amount = investAmount.mulWadDown(allocationPercent);
+                // first check if allocation Percent is greater than zero
+                (flashLoanAmount, isSupplyBorrow) = _calcRebalanceFlashLoanAmount(adapter, amount);
+
+                rebalanceDataParams.push(RebalanceDataParams(adapter, amount, flashLoanAmount, isSupplyBorrow));
+            } else {
+                uint256 ltv = getLtv(adapter);
+                // even if there is no allocation here, we still want to check if a disinvest is needed in this protocol
+                if (ltv > targetLtv[adapter] + disinvestThreshold) {
+                    (flashLoanAmount, isSupplyBorrow) = _calcRebalanceFlashLoanAmount(adapter, 0);
+
+                    rebalanceDataParams.push(RebalanceDataParams(adapter, 0, flashLoanAmount, false));
+                }
+            }
+        }
+    }
+
+    function _rebalance() internal {
+        _updateRebalanceParams();
+        (bytes[] memory callData, uint256 totalFlashLoanAmount) = _getRebalanceCallData();
+
+        if (totalFlashLoanAmount > minimumInvestAmount.mulWadDown(getLeverage())) {
+            console2.log("---- Running Rebalance ----");
+            vault.rebalance(_investAmount(), totalFlashLoanAmount, callData);
+        }
+    }
+
+    function _getRebalanceCallData() internal returns (bytes[] memory, uint256) {
+        uint256 stEthRateTolerance = 0.9995e18;
+        uint256 n = rebalanceDataParams.length;
+
+        RebalanceDataParams memory data;
+
+        bytes[] memory temp = new bytes[](n);
```

Comment:
try using "DoubleEndedQueue" library to eliminate this temp variable since it allows pushing items from the front and back. So you can fill up this struct and just convert it to `bytes[]` before the call to reinvest

Reply:
Great idea. I will try that.

Reply:
I am not sure that will work, since DoubleEndedQueue expects a bytes32 object but here we are adding bytes object.

Reply:
i think we can work around that by simply converting bytes32 -> bytes and bytes -> bytes32.

Reply:
ok, nvm it's not so simple :)